### ISSUE TITLE
wrong code at -O1 and above on x86_64-linux-gnu (regression from 12.0.1)

### ISSUE THREAD

**Issue by zhendongsu** [llvm/llvm-project#52722]
It also affects 13.0.0, but not 12.0.1.

```
[506] % clangtk -v
clang version 14.0.0 (https://github.com/llvm/llvm-project.git 9912bed7306f0157b6425f233eec56d04bb536cf)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
[507] % 
[507] % clangtk -O0 small.c; ./a.out
[508] % 
[508] % clangtk -O1 small.c
[509] % ./a.out
Floating point exception
[510] % 
[510] % cat small.c
int a, b, c, *d = &c, e, f;
int main() {
  int g[2] = {0, 0};
  while (1) {
    int *h;
    for (e = 0; e < 2; e++)
      h = &g[e];
    if (h != g || !h)
      while (1) {
        a = ~a;
        f = ~0 % a;
        break;
      }
    for (; b < 2; b++)
      if (!(h && *d)) {
        *g = 0;
        return 0;
      }
  }
  return 0;
}
```

**Comment by zhendongsu** [comment]
From Compiler Explorer: https://godbolt.org/z/Yo6GKTb4P

**Comment by fhahn** [comment]
It looks like this has also been fixed by 4a699ae9c6a6fde93ec4714d79025df58eb62a72. It fails with the change reverted, but passes on current `main` and on the Godbolt link above.
